### ISSUE TITLE
Smarter &plist, &alist destructuring

### ISSUE THREAD

**Issue by Fuco1** [magnars/dash.el#111]
Most of the time we write the destructure like this:

```
(&plist :key key :otherkey otherkey)
```

where the variable to which we destructure is the same as they key sans the `:`. I'm proposing this: when `*` is appended to the selector (so it's e.g. `&plist*`) the name of the binding variable is derived from the key in the following "smart way":
- a key `:foo` is converted into `foo` variable
- a key `foo` is converted into `foo` variable
- a key `"foo"` is converted into `foo` variable

This covers pretty much 99% of use-cases of mapping structures.

Ideas?  Instead of `*` we might use a different symbol, but stars often means some variation in lisp so I think it fits well.

**Comment by fommil** [comment]
for `plist`s, `&key` would be consistent with `destructuring-bind` in common lisp http://www.gigamonkeys.com/book/beyond-lists-other-uses-for-cons-cells.html

**Comment by Fuco1** [comment]
Right now `&keys` is used for "trailing plist", similar to `cl-defun`, so you can do `(1 2 :a a :b b :c c)` as `(-let (((one two &keys :b b) ...))`, so using `&key` would probably cause a bit of confusion.

From the link, the `&whole` keyword is here called `&as`, modeled after clojure (in general, we lean more on clojure than CL).

**Comment by fommil** [comment]
I only learnt about `destructuring-bind` myself recently (I'm reading the linked book, it's very good), the limitation here is that the following fails because there are parts of the expr that are not matched

```
  (cl-destructuring-bind 
      (&key a b) 
      '(:a "foo" :b 13 :c "bar") 
    (list a b)) 
```

and although it is wordier, it does have the advantage over `-let` that there are less nested pairs of brackets.

I agree with the `&keys` vs `&key` potential for confusion but I still wouldn't rule it out. It might just need a note beside the `&keys` documentation (which seems like a pretty rare usecase maybe?) that it is not to be confused with the common-lisp style `&key`.

**Comment by fommil** [comment]
http://emacs.stackexchange.com/questions/22542/cl-destructuring-bind-on-partial-plist/22550#22550 it can be done

OK, I'll go back to preferring `&plist*` to avoid confusion with keys and the common lisp syntax.

**Comment by fommil** [comment]
I discovered this recently, for inspiration https://elpa.gnu.org/packages/let-alist.html

**Comment by Fuco1** [comment]
I don't like the leading dot syntax very much, but otherwise it's pretty slick, especially the "automagic" nesting. @nicferrier has a similar thing in `kv` package.

We're getting a lot of destructuring going on, I'm not sure which package is going to win :D  It would be best to get some sort of a standard going on in "elisp" itself.  But pcase is not an option, it's terrible /o\.

**Comment by Fuco1** [comment]
Reminds me of https://github.com/stedolan/jq syntax a bit.  Speaking of json... json destructuring would be damn rad as well... `&json`?

**Comment by fommil** [comment]
FYI

```elisp
(defmacro plist-bind (args expr &rest body)
  ;; http://emacs.stackexchange.com/questions/22542
  "`destructuring-bind' without the boilerplate for plists."
  `(cl-destructuring-bind
       (&key ,@args &allow-other-keys)
       ,expr
     ,@body))
```

**Comment by alphapapa** [comment]
Having recently (re?)discovered the usefulness of `-let`, I would really love to see this implemented.  The boilerplate for destructuring plists is so repetitive:

```el
(let ((pl (list (list :two 2 :one 1))))
  (cl-loop for l in pl
           collect (-let (((plist &as &plist :two two :one one) l))
                     (list one two)))) => ((1 2))
```

This is bad enough, but for plists with more keys that have longer names, it gets...messy.  It would be a relief to be able to do:

```el
(let ((pl (list (list :two 2 :one 1))))
  (cl-loop for l in pl
           collect (-let (((&plist*) l))
                     (list one two)))) => ((1 2))
```

My elisp-fu is relatively weak, but is there anything I can do to help with this?  Thanks.

**Comment by Fuco1** [comment]
So you propose that in addition to the OP we also bind *everything* if the `&plist*` selector is empty?  I mean, why not, but that wouldn't work at compile time so there would be some trickery with the macro.

By the way, you don't have to use `plist &as`, but I guess you knew that :P

These days, dash is getting a bit behind and people are slowly droping it in favour of seq and the built-ins (emacs 25 doesn't suck as much as 23 did 5 years ago).  I might still do this though because I prefer dash's API (for obvious reasons).  Thanks for bringing this up, might motivate me to finish through with the work.

**Comment by alphapapa** [comment]
> So you propose that in addition to the OP we also bind everything if the &plist* selector is empty? I mean, why not, but that wouldn't work at compile time so there would be some trickery with the macro.

I'm not sure exactly what you mean.  I'm not asking that the whole plist be bound, just the elements of it, like in the second example I gave.  I just tire of typing each key 3 times.  :)

> By the way, you don't have to use plist &as, but I guess you knew that :P

Haven't really used that before, so I just went by the docstring:

```
You can name the source using the syntax SYMBOL &as PATTERN.
This syntax works with lists (proper or improper), vectors and
all types of maps.

 ...

  (plist &as &plist :b b) (list :a 1 :b 2 :c 3)
```

> These days, dash is getting a bit behind and people are slowly droping it in favour of seq and the built-ins (emacs 25 doesn't suck as much as 23 did 5 years ago). 

I like dash a lot.  Please keep up the great work on it.  :)

**Comment by Fuco1** [comment]
In the current syntax, you can write

    ((plist &as &plist :two two :one one) l)

as just

    ((&plist :two two :one one) l)

and the proposition in OP would turn this into

    ((&plist* :two :one) l)

with `one` and `two` automatically generated.

I assumed that by `((&plist*) l)` you ment to create bindings for all the keys in the plist, which seems neat but I'm not yet sure how to achieve that (because you can't generate code at macro-expansion time as you have no idea what is in `l`.

**Comment by alphapapa** [comment]
I see, thanks for clarifying.  Well, yes, ideally the last part would be great, but even if that's not possible, the OP's idea would also be a big improvement.  :)

**Comment by wasamasa** [comment]
I've recently tried out `&alist` inside `-let` and wished for a way to just have to specify keys.  To avoid ambiguity/smartness I thought of porting Clojure's convention of indicating the specific type as they do by having `:keys [foo bar]` (or `:syms` or `:strs`) in the destructuring directive.  A possible syntax for this would be `&alist-syms`, `&plist-keys`, `&hash-strs` and all the other possible permutations.  What do you think?

**Comment by Fuco1** [comment]
@wasamasa So how would that differ from `&alist*` other then the naming?  Or are you just suggesting a better name for the symbol (which I'm not opposed against).  If there is some change in usage then please an example :)

If anyone wants to hack on this please go ahead, I'm rather bummed right out about a lot of things and not really in a mood to depress myself further with Emacs :P

**Comment by wasamasa** [comment]
@Fuco1 `&alist*` would be the smart version whereas I'm proposing explicit ones that only work for the specified type and would otherwise fail.

**Comment by Fuco1** [comment]
So if I understand this correctly, if you say `:keys [foo bar]` it will only destructure on the keys which are `:foo` and `:bar` and put it in the variables `foo` and `bar`, if you do `:strs [foo bar]` it will destructure on keys which are `"foo"` and `"bar"`.  Is that correct?

This seems like a good addition probably, a little bit of type sanity.

**Comment by alphapapa** [comment]
Why have separate keywords for strings or symbols when it could be, e.g. `:keys [foo bar "baz"]`?

**Comment by Fuco1** [comment]
Then I don't see how is it different from the original proposal.

**Comment by wasamasa** [comment]
@alphapapa `:keys` works for Clojure because they only have one map type whereas in elisp we've got to deal with hash tables, alists and plists.  The existing solution for this is specifying the type of the thing to destructure, what I've proposed is appending the type of the keys to work on to that type.

**Comment by alphapapa** [comment]
Sorry, I still don't understand.  Why is it necessary to specify `:keys` or `:syms` when, e.g. `[foo bar "baz"]` can be inferred as two symbols and a string?

**Comment by Fuco1** [comment]
Yes I also don't understand it anymore (I thought I did), but the original proposal already only works on the specified keys (you specify keys and the variables get inferred, not the other way).

**Comment by Fuco1** [comment]
On `feature/smart-kv-destructuring` there is now an attempt to make this smart but predictable without any special syntax.

tl;dr if the thing following the `:keyword`, `'symbol` or `"string"` is something that is not a valid match-form i.e. (symbol, cons or vector) the match-form is derived as per the rules in OP.

So you can write

```elisp
(-let (((&plist :foo :bar) (list :foo 1 :bar 2))
       ((&plist :foo2 foo2 :bar2) (list :foo2 1 :bar2 2))
       ((&plist :foo3 :bar3 bar3) (list :foo3 1 :bar3 2))
       ((&plist :foo4 foo4 :bar4 bar4) (list :foo4 1 :bar4 2)))
  (list foo bar foo2 bar2 foo3 bar3 foo4 bar4))
```

expands to

```elisp
(let
    ((input0
      (list :foo 1 :bar 2))
     (input1
      (list :foo2 1 :bar2 2))
     (input2
      (list :foo3 1 :bar3 2))
     (input3
      (list :foo4 1 :bar4 2)))
  (let*
      ((foo
        (plist-get input0 :foo))
       (bar
        (plist-get input0 :bar))
       (foo2
        (plist-get input1 :foo2))
       (bar2
        (plist-get input1 :bar2))
       (foo3
        (plist-get input2 :foo3))
       (bar3
        (plist-get input2 :bar3))
       (foo4
        (plist-get input3 :foo4))
       (bar4
        (plist-get input3 :bar4)))
    (list foo bar foo2 bar2 foo3 bar3 foo4 bar4)))
```

**Comment by alphapapa** [comment]
That looks great, Matus!  It's the perfect way to do it: infer the variable names by default, but allow overriding when necessary.  I look forward to using that!

**Comment by Fuco1** [comment]
Yea, I got inspired a bit by how javascript does it, you can write e.g. `var {foo, bar} = object` or `var {foo: x, bar: y} = object`... so I think it makes a lot of sense.

The best thing is that it is possible to do static analysis so if this ever fails it fails when you compile your package and so people shouldn't suffer any nonsense at runtime.

**Comment by fommil** [comment]
thank you so much, I am very happy that you took the time to do this! :grinning:

**Comment by Fuco1** [comment]
@fommil I'm also quite happy :D It was long overdue.
